crypto/property/property.c

```diff
@@ -0,0 +1,507 @@
+} IMPL_CACHE_FLUSH;
+
+DEFINE_SPARSE_ARRAY_OF(ALGORITHM);
+
+static OSSL_IMPL_STORE *g_implementations = NULL;
+
+static void ossl_impl_cache_flush(OSSL_IMPL_STORE *impls, int nid);
+static void ossl_impl_cache_flush_all(OSSL_IMPL_STORE *c);
+
+int ossl_property_read_lock(OSSL_IMPL_STORE *p)
+{
+    if (p == NULL)
+        p = g_implementations;
+    return CRYPTO_THREAD_read_lock(p->lock);
+}
+
+int ossl_property_write_lock(OSSL_IMPL_STORE *p)
+{
+    if (p == NULL)
+        p = g_implementations;
+    return CRYPTO_THREAD_write_lock(p->lock);
+}
+
+int ossl_property_unlock(OSSL_IMPL_STORE *p)
+{
+    if (p == NULL)
+        p = g_implementations;
+    return CRYPTO_THREAD_unlock(p->lock);
+}
+
+static unsigned long query_hash(const QUERY *a)
+{
+    return OPENSSL_LH_strhash(a->query);
+}
+
+static int query_cmp(const QUERY *a, const QUERY *b)
+{
+    return strcmp(a->query, b->query);
+}
+
+static void impl_free(IMPLEMENTATION *impl)
+{
+    if (impl != NULL) {
+        if (impl->implementation_destruct)
+            impl->implementation_destruct(impl->implementation);
+        OPENSSL_free(impl);
+    }
+}
+
+static void impl_cache_free(QUERY *elem)
+{
+    OPENSSL_free(elem);
+}
+
+static void alg_cleanup(ALGORITHM *a)
+{
+    if (a != NULL) {
+        sk_IMPLEMENTATION_pop_free(a->impls, &impl_free);
+        lh_QUERY_doall(a->cache, &impl_cache_free);
+        lh_QUERY_free(a->cache);
+        OPENSSL_free(a);
+    }
+}
+
+OSSL_IMPL_STORE *ossl_impl_store_new(void)
+{
+    OSSL_IMPL_STORE *res = OPENSSL_zalloc(sizeof(*res));
+
+    if (res != NULL) {
+        if ((res->algs = ossl_sa_ALGORITHM_new()) == NULL) {
+            OPENSSL_free(res);
+            return NULL;
+        }
+        if ((res->lock = CRYPTO_THREAD_lock_new()) == NULL) {
+            OPENSSL_free(res->algs);
+            OPENSSL_free(res);
+            return NULL;
+        }
+    }
+    return res;
+}
+
+void ossl_impl_store_free(OSSL_IMPL_STORE *impls)
+{
+    ossl_sa_ALGORITHM_doall(impls->algs, &alg_cleanup);
+    ossl_sa_ALGORITHM_free(impls->algs);
+    ossl_property_free(impls->global_properties);
+    CRYPTO_THREAD_lock_free(impls->lock);
+    OPENSSL_free(impls);
+}
+
+static ALGORITHM *ossl_impl_store_retrieve(OSSL_IMPL_STORE *impls, int nid)
+{
+    return ossl_sa_ALGORITHM_get(impls->algs, nid);
+}
+
+static int ossl_impl_store_insert(OSSL_IMPL_STORE *impls, ALGORITHM *alg)
+{
+        return ossl_sa_ALGORITHM_set(impls->algs, alg->nid, alg);
+}
+
+int ossl_impl_store_init(void)
+{
+    g_implementations = ossl_impl_store_new();
+    if (ossl_property_string_init()
+            && ossl_prop_defn_init()
+            && ossl_property_parse_init()
+            && g_implementations != NULL)
+        return 1;
+
+    ossl_impl_store_cleanup();
+    return 0;
+}
+
+void ossl_impl_store_cleanup(void)
+{
+    ossl_property_string_cleanup();
+    ossl_prop_defn_cleanup();
+    ossl_impl_store_free(g_implementations);
+    g_implementations = NULL;
+}
```

Comment:
Please make sure that this is called automatically, either as part of the cleanup process, or in here with the help of `OPENSSL_atexit()` (see #8225 to see how I do that)